### ISSUE TITLE
MultiProcessRunner RuntimeError

### ISSUE THREAD

**Issue by dps42** [tensorpack/tensorpack#1219]
If you're asking about an unexpected problem which you do not know the root cause,
use this template. __PLEASE DO NOT DELETE THIS TEMPLATE, FILL IT__:

If you already know the root cause to your problem,
feel free to delete everything in this template.

### 1. What you did:

(1) **If you're using examples, what's the command you run:**

(2) **If you're using examples, have you made any changes to the examples? Paste `git status; git diff` here:**

(3) **If not using examples, tell us what you did:**

  It's always better to copy-paste what you did than to describe them.

  Please try to provide enough information to let other __reproduce__ your issues. 
  Without reproducing the issue, we may not be able to investigate it.

  I tried to follow the "Efficient Dataflow" tutorial, continuing from https://github.com/tensorpack/tensorpack/issues/1209. 

  
### 2. What you observed:

(1) **Include the ENTIRE logs here:**

It's always better to copy-paste what you observed instead of describing them.

It's always better to paste **as much as possible**, although sometimes a partial log is OK.

Tensorpack typically saves stdout to its training log.
If stderr is relevant, you can run a command with `my_command 2>&1 | tee logs.txt`
to save both stdout and stderr to one file.

[0528 10:55:08 @parallel.py:195] WRN MultiProcessRunner does support Windows. However, Windows requires more strict picklability on processes, which may lead of failure on some of the code.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\spawn.py", line 106, in spawn_main
    exitcode = _main(fd)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\spawn.py", line 115, in _main
    prepare(preparation_data)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\spawn.py", line 226, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\spawn.py", line 278, in _fixup_main_from_path
    run_name="__mp_main__")
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\runpy.py", line 254, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\AI_Workspace\z_debug\load_lmdb.py", line 79, in <module>
    load_lmdb3()
  File "C:\AI_Workspace\z_debug\load_lmdb.py", line 69, in load_lmdb3
    ds = MultiProcessRunner(ds, 5000, 1)       # NOTE: PrefetchData() deprecated in May 2019
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\site-packages\tensorpack\dataflow\parallel.py", line 214, in __init__
    start_proc_mask_signal(self.procs)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\site-packages\tensorpack\utils\concurrency.py", line 244, in start_proc_mask_signal
    p.start()
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\site-packages\tensorpack\utils\concurrency.py", line 216, in mask_sigint
    yield True
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\site-packages\tensorpack\utils\concurrency.py", line 244, in start_proc_mask_signal
    p.start()
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\process.py", line 105, in start
    self._popen = self._Popen(self)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\context.py", line 313, in _Popen
    return Popen(process_obj)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\popen_spawn_win32.py", line 34, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\spawn.py", line 144, in get_preparation_data
    _check_not_importing_main()
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\spawn.py", line 137, in _check_not_importing_main
    is not going to be frozen to produce an executable.''')
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.

I will attach the code here: 
[z_debug.zip](https://github.com/tensorpack/tensorpack/files/3228301/z_debug.zip)

But please notice that the LMDB file I'm using is too large to be attached to the zip file.
The LMDB file was created from the same "debug2.py" but with more images and data entries.

From load_lmdb3() function, the code crashed with "MultiProcessRunner()" with a RuntimeError.
Maybe another Windows issue ?
I had the same error before PrefetchData() was renamed to MultiProcessRunner()

(2) **Other observations, if any:**
For example, CPU/GPU utilization, output images, tensorboard curves, if relevant to your issue.

### 3. What you expected, if not obvious.

If you expect higher speed, please read
http://tensorpack.readthedocs.io/tutorial/performance-tuning.html 
before posting.

If you expect certain accuracy, only in one of the two conditions can we help with it:
(1) You're unable to reproduce the accuracy documented in tensorpack examples.
(2) It appears to be a tensorpack bug.

Otherwise, how to train a model to certain accuracy is a machine learning question.
We do not answer machine learning questions and it is your responsibility to 
figure out how to make your models more accurate.

### 4. Your environment:
  + Paste the output of this command: `python -c 'import tensorpack.tfutils as u; print(u.collect_env_info())'`
    If this command failed, tell us your version of Python/TF/tensorpack.
  + You can install Tensorpack master by `pip install -U git+https://github.com/ppwwyyxx/tensorpack.git`
    and see if your issue is already solved.
  + If you're not using tensorpack under a normal command line shell (e.g.,
    using an IDE or jupyter notebook), please retry under a normal command line shell. 
  + Include relevant hardware information, e.g. number of GPUs used for training, amount of RAM.

You may often want to provide extra information related to your issue, but
at the minimum please try to provide the above information __accurately__ to save effort in the investigation.

 Windows 10. I think no GPU was used at the moment.

**Comment by ppwwyyxx** [comment]
Your log has already told you the reason:
```
This probably means that you are not using fork to start your
    child processes and you have forgotten to use the proper idiom
    in the main module:

        if __name__ == '__main__':
            freeze_support()
            ...

    The "freeze_support()" line can be omitted if the program
    is not going to be frozen to produce an executable.
```

**Comment by dps42** [comment]
Hi. I changed the code to run the function in "if __name__ == '__main__': but the issue still exists. Here's the error message:

[0528 11:44:49 @parallel.py:195] WRN MultiProcessRunner does support Windows. However, Windows requires more strict picklability on processes, which may lead of failure on some of the code.
Traceback (most recent call last):
  File "load_lmdb.py", line 80, in <module>
    load_lmdb3()
  File "load_lmdb.py", line 69, in load_lmdb3
    ds = MultiProcessRunner(ds, 5000, 1)       # NOTE: PrefetchData() deprecated in May 2019
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\site-packages\tensorpack\dataflow\parallel.py", line 214, in __init__
    start_proc_mask_signal(self.procs)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\site-packages\tensorpack\utils\concurrency.py", line 244, in start_proc_mask_signal
    p.start()
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\process.py", line 105, in start
    self._popen = self._Popen(self)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\context.py", line 313, in _Popen
    return Popen(process_obj)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\popen_spawn_win32.py", line 66, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <class 'Environment'>: attribute lookup Environment on builtins failed

(dps42_dev) C:\AI_Workspace\z_debug>Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\spawn.py", line 106, in spawn_main
    exitcode = _main(fd)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\spawn.py", line 116, in _main
    self = pickle.load(from_parent)
EOFError: Ran out of input

**Comment by ppwwyyxx** [comment]
That's a different issue.

> _pickle.PicklingError: Can't pickle <class 'Environment'>: attribute lookup Environment on builtins failed

Your class cannot be serialized by Python. You need to make it serializable by Python's pickle.

**Comment by dps42** [comment]
I don't understand what you mean, can you provide some more details?? And why is pickle even relevant here?? I'm just trying to load from an LMDB file. load_lmdb2() with BatchData() just worked fine. I start to have an issue when MultiProcessRunner() comes in.

**Comment by ppwwyyxx** [comment]
It's required by python's multiprocessing on windows, that your class needs to be pickled to be run in parallel.

**Comment by dps42** [comment]
So do you have a suggestion for how to resolve the issue then? I don't have a class named Environment, and I'm not sure what's causing the issue.

**Comment by ppwwyyxx** [comment]
That's interesting. Perhaps the dataflow you use refers to some other objects or functions in third-party libraries which involve a class named "Environment".

**Comment by ppwwyyxx** [comment]
Looks like the "Environment" class is part of the lmdb database. So I guess you cannot use multiprocessing with LMDBSerializer.load on windows.

**Comment by ppwwyyxx** [comment]
```python
class MyOwnLMDBData(DataFlow):
    def __init__(self, path):
        self.path = path
    def __iter__(self):
        d = LMDBData(self.path)
        yield from d
```
Using this instead of `LDMBSerializer.load` might makes it possible to run under multiprocessing, but perhaps less efficient than it should be.

**Comment by dps42** [comment]
thanks for the suggestion. I'll try it later and see if it works.

**Comment by ppwwyyxx** [comment]
Sorry the correct code should be:
```python
class MyOwnLMDBData(DataFlow):
    def __init__(self, path):
        self.path = path
    def __iter__(self):
        d = LMDBSerializer.load(self.path)
        d.reset_state()
        yield from d
```

**Comment by ppwwyyxx** [comment]
The commit 0a0b387e1f5993e7c7e1f09f29dd79155c2ea03c on LMDBData should make it usable on windows with multiprocessing!

**Comment by dps42** [comment]
did you mean it should work without the code snippet you suggested above?

**Comment by ppwwyyxx** [comment]
Yes. After the commit, `LMDBSerializer.load` should work on windows with multiprocessing.

**Comment by dps42** [comment]
If I don't use the code snippet above, I have another issue: 
[0528 14:04:13 @parallel.py:195] WRN MultiProcessRunner does support Windows. However, Windows requires more strict picklability on processes, which may lead of failure on some of the code.
Traceback (most recent call last):
  File "load_lmdb.py", line 80, in <module>
    load_lmdb3()
  File "load_lmdb.py", line 73, in load_lmdb3
    ds = MultiProcessRunner(ds, 25)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\site-packages\tensorpack\dataflow\parallel.py", line 201, in __init__
    assert num_proc > 0, num_proc
TypeError: unorderable types: NoneType() > int()

Before trying that above,  I had a few typos. Sorry. fixed below. 
def load_lmdb3():
    ds = LMDBSerializer.load( LMDB_PATH, shuffle=False)
    
    num_images = len(ds)
    print( num_images )
    
    ds = LocallyShuffleData(ds, 50000)
    ds = MultiProcessRunner(ds, 5000, 1)       # NOTE: PrefetchData() deprecated in May 2019
    ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
    # ds = AugmentImageComponent( ds, lots_of_augmentors)
    ds = AugmentImageComponent( ds, [imaug.Resize(224)] )
    ds = MultiProcessRunner(ds, 25) 
    ds = BatchData(ds, 256, use_list=True)
    
    TestDataSpeed(ds).start()

**Comment by dps42** [comment]
wait. sorry. I'll try to insert another argument to MultiProcessRunner().

**Comment by dps42** [comment]
yes. I tried each of the combinations below instead of "ds = MultiProcessRunner(ds, 25) "
ds = MultiProcessRunner(ds, 25,1)
ds = MultiProcessRunner(ds, 25,2)
ds = MultiProcessRunner(ds, 5000,1)   

But I seem to have the same issue...

**Comment by ppwwyyxx** [comment]
You have two `MultiProcessRunner` to change.

**Comment by dps42** [comment]
ds = MultiProcessRunner(ds, 5000, 1)  already had two arguments.

**Comment by ppwwyyxx** [comment]
You have two MultiProcessRunner in your code

**Comment by dps42** [comment]
I directly copied the example code from "Efficient Dataflow" tutorial.

ds = LMDBSerializer.load(db, shuffle=False)
ds = LocallyShuffleData(ds, 50000)
ds = MultiProcessRunner(ds, 5000, 1)
ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
ds = AugmentImageComponent(ds, lots_of_augmentors)
ds = MultiProcessRunnerZMQ(ds, 25)
ds = BatchData(ds, 256)

And I replaced MultiProcessRunnerZMQ with MultiProcessRunner. 
Did you mean I need only one MultiProcessRunner()?

**Comment by ppwwyyxx** [comment]
It is very hard to understand the conversation. If this is your code:
> ds = LMDBSerializer.load(db, shuffle=False)
ds = LocallyShuffleData(ds, 50000)
ds = MultiProcessRunner(ds, 5000, 1)
ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
ds = AugmentImageComponent(ds, lots_of_augmentors)
ds = MultiProcessRunnerZMQ(ds, 25)
ds = BatchData(ds, 256)



Could you post again what is the error?

**Comment by dps42** [comment]
As posted above, my code was 

def load_lmdb3():
ds = LMDBSerializer.load( LMDB_PATH, shuffle=False)
ds = LocallyShuffleData(ds, 50000)
ds = MultiProcessRunner(ds, 5000, 1)       # NOTE: PrefetchData() deprecated in May 2019
ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
def load_lmdb3():
ds = LMDBSerializer.load( LMDB_PATH, shuffle=False)

num_images = len(ds)
print( num_images )

ds = LocallyShuffleData(ds, 50000)
ds = MultiProcessRunner(ds, 5000, 1)       # NOTE: PrefetchData() deprecated in May 2019
ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
# ds = AugmentImageComponent( ds, lots_of_augmentors)
ds = AugmentImageComponent( ds, [imaug.Resize(224)] )
ds = MultiProcessRunner(ds, 25) 
ds = BatchData(ds, 256, use_list=True)

TestDataSpeed(ds).start()

I replaced MultiProcessRunnerZMQ() with MultiProcessRunner, because ZMQ is not available on Windows.  But as mentioned above, I faced another issue:

[0528 14:04:13 @parallel.py:195] WRN MultiProcessRunner does support Windows. However, Windows requires more strict picklability on processes, which may lead of failure on some of the code.
Traceback (most recent call last):
File "load_lmdb.py", line 80, in 
load_lmdb3()
File "load_lmdb.py", line 73, in load_lmdb3
ds = MultiProcessRunner(ds, 25)
File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\site-packages\tensorpack\dataflow\parallel.py", line 201, in init
assert num_proc > 0, num_proc
TypeError: unorderable types: NoneType() > int()

Thank you for the patience and support.

**Comment by ppwwyyxx** [comment]
> File "load_lmdb.py", line 73, in load_lmdb3
ds = MultiProcessRunner(ds, 25)

As shown in the error log, you missed an argument here.

**Comment by dps42** [comment]
From the previous conversation, I reported that I'd tried each of the three combinations instead of ds = MultiProcessRunner(ds, 25) below:
```
ds = MultiProcessRunner(ds, 25,1)
ds = MultiProcessRunner(ds, 25,2)
ds = MultiProcessRunner(ds, 5000,1)
```
But I had the same error log. Is there any missing argument?

**Comment by ppwwyyxx** [comment]
The log clearly says that 
> File "load_lmdb.py", line 73, in load_lmdb3
ds = MultiProcessRunner(ds, 25)

which means at line 73 in `load_lmdb.py`, you have `ds = MultiProcessRunner(ds, 25)` which is wrong and  should be changed. If you do change it, you will not see this error.

**Comment by dps42** [comment]
Sorry. I copied it from the other terminal window. I apologize. 

I updated Tensorpack and tried the new code with the additional argument, but it seems that the previous problem still exists. 

[0528 15:00:32 @parallel.py:195] WRN MultiProcessRunner does support Windows. However, Windows requires more strict picklability on processes, which may lead of failure on some of the code.
Traceback (most recent call last):
  File "load_lmdb.py", line 80, in <module>
    load_lmdb3()
  File "load_lmdb.py", line 73, in load_lmdb3
    ds = MultiProcessRunner(ds, 25,1)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\site-packages\tensorpack\dataflow\parallel.py", line 214, in __init__
    start_proc_mask_signal(self.procs)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\site-packages\tensorpack\utils\concurrency.py", line 244, in start_proc_mask_signal
    p.start()
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\process.py", line 105, in start
    self._popen = self._Popen(self)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\context.py", line 313, in _Popen
    return Popen(process_obj)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\popen_spawn_win32.py", line 66, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'load_lmdb3.<locals>.<lambda>'

(dps42_dev) C:\AI_Workspace\z_debug>Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\spawn.py", line 106, in spawn_main
    exitcode = _main(fd)
  File "C:\Users\dps42\AppData\Local\Continuum\miniconda3\envs\dps42_dev\lib\multiprocessing\spawn.py", line 116, in _main
    self = pickle.load(from_parent)
EOFError: Ran out of input

**Comment by ppwwyyxx** [comment]
This is a different problem. That means you should not use a lambda function in `load_lmdb3` (because it's not pickleable and it's running on windows). You should write a new function instead of the lambda function.

**Comment by dps42** [comment]
Hi. The only lambda I'm seeing is one line from the example code from "Efficient Dataflow":
``` ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
```
The following code worked, without a lambda function as you suggested.

```
def decode_function(x):
    return cv2.imdecode(x, cv2.IMREAD_COLOR)

def load_lmdb3():
    ds = LMDBSerializer.load( LMDB_PATH, shuffle=False)
    
    num_images = len(ds)
    print( num_images )
    
    ds = LocallyShuffleData(ds, 50000)
    ds = MultiProcessRunner(ds, 5000, 2)       # NOTE: PrefetchData() deprecated in May 2019
    ds = MapDataComponent(ds, decode_function, 0)
    # ds = AugmentImageComponent( ds, lots_of_augmentors)
    ds = AugmentImageComponent( ds, [imgaug.Resize(224)] )  # temporary, will replace
    ds = MultiProcessRunner(ds, 25,2) 
    ds = BatchData(ds, 256, use_list=True)
    
    TestDataSpeed(ds).start()

# load LMDB
if __name__ == '__main__':
    load_lmdb3()    
```
Thank you so much for the patience.

**Comment by ppwwyyxx** [comment]
Great. I've added these common issues in the efficient dataflow tutorial.
